File: code/ch07-01/calculus.erl
%% @author J D Eisenberg <jdavid.eisenberg@gmail.com>
%% @doc Find the derivative of a function Fn at point X.
%% @copyright 2013 J D Eisenberg
%% @version 0.1

-module(calculus).
-export([derivative/2]).

%% @doc Calculate derivative by classical definition.
%% (Fn(X + H) - Fn(X)) / H

-spec(derivative(function(), float()) -> float()).

derivative(Fn, X) ->
	Delta = 1.0e-10,
	(Fn(X + Delta) - Fn(X)) / Delta.

File: code/ch08-01/calculus.ex
defmodule Calculus do
  @moduledoc """ 
  from *Études for Elixir*, O'Reilly Media, Inc., 2013.
  Copyright 2013 by J. David Eisenberg.
  """ 
  @vsn 0.1 

  @doc """
  Calculate the approximation to the derivative of
  function f and point x by using the mathematical
  definition of a derivative.
  """
  @spec derivative(fun, number) :: number
  
  def derivative(f, x) do
    delta = 1.0e-10
    (f.(x + delta) - f.(x)) / delta
  end
end

